src/sentry/api/endpoints/organization_dashboard_details.py

```diff
@@ -1,11 +1,97 @@
 from __future__ import absolute_import
 
+from django.db.models import Max
+from rest_framework import serializers
+from rest_framework.response import Response
+
 from sentry.api.base import DocSection
 from sentry.api.bases.dashboard import (
     OrganizationDashboardEndpoint
 )
 from sentry.api.serializers import serialize
-from sentry.models import ObjectStatus
+from sentry.api.serializers.rest_framework import ListField, ValidationError
+from sentry.models import ObjectStatus, Widget
+
+
+def get_next_dashboard_order(dashboard_id):
+    max_order = Widget.objects.filter(
+        dashboard_id=dashboard_id,
+    ).aggregate(Max('order'))
+    return max_order['order__max'] + 1
+
+
+def remove_widgets(dashboard_widgets, widget_data):
+    """
+    Removes current widgets belonging to dashboard not in widget_data.
+    Returns remaining widgets.
+    """
+    widget_titles = [wd['id'] for wd in widget_data]
+    dashboard_widgets.exclude(
+        id__in=widget_titles
+    ).update(status=ObjectStatus.PENDING_DELETION)
+
+    return dashboard_widgets.filter(
+        id__in=widget_titles
+    )
+
+
+def reorder_widgets(dashboard_id, widget_data):
+    """
+    Reorders Widgets given the relative order desired,
+    reorders widgets in the next possible set of numbers
+    i.e if order of widgets is 1, 2, 3
+        the reordered widgets will have order 4, 5, 6
+    """
+    dashboard_widgets = Widget.objects.filter(
+        dashboard_id=dashboard_id,
+    )
+    dashboard_widgets = list(remove_widgets(dashboard_widgets, widget_data))
+
+    # dashboard_widgets and widget_data should now have the same widgets
+    widget_data.sort(key=lambda x: x['order'])
+
+    next_order = get_next_dashboard_order(dashboard_id)
+    for index, data in enumerate(widget_data):
+        for widget in dashboard_widgets:
+            if widget.id == data['id']:
+                widget.order = next_order + index
+                widget.save()
+                break
+
+
+class WidgetSerializer(serializers.Serializer):
+    order = serializers.IntegerField(min_value=0, required=True)
```

Comment:
Do you want an explicit order? How are sparse order values going to be handled?

Reply:
I would just assume they would be relative ordering. So if you have:
```
(id=1, order=2), (id=2, order=5), (id=3, order=10)
```
And you want to switch widgets with id's 1 and 2.
The user would give:
```
{
 widgets: [{id:2, order:1}, {id:1, order:2}, {id:3, order:3}
}
```
I'm looking at their relative order. And so the result in the database is:
```
(id=1, order=12), (id=2, order=11), (id=3, order=13)
```

I think the reason why I ended up doing it this way was because I couldn't think of an efficient enough way to swap numbers as they should be unique. An alternative method I thought of if we really wanted to keep the same numbers would be to delete one (or add a place holder) as one does when swapping numbers.

If you feel that would be better do let me know.

Reply:
JFYI, you can actually do this in Postgres with a deferred constraint, which basically allows you to defer constraint checking until after a transaction commits. Unfortunately Mysql and sqlite don't support this, supporting multiple databases is a pain :(

Reply:
@markstory let me know what you think of this approach.